src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,499 @@ namespace System
+        {
+            if (trimElements.Length <= 1) // Optimize for N > 1
+            {
+                return trimElements.Length == 0 ? span : TrimStart(span, trimElements[0]);
+            }
+
+            int start = ClampStart(span, trimElements);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            if (trimElements.Length <= 1) // Optimize for N > 1
+            {
+                return trimElements.Length == 0 ? span : TrimEnd(span, trimElements[0]);
+            }
+
+            int length = ClampEnd(span, 0, trimElements);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        private static int ClampStart<T>(ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            int start = 0;
+
+            for (; start < span.Length; start++)
+            {
+                if (!SequentialContains(trimElements, span[start]))
+                    break;
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="start">The start index from which to being searching.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        private static int ClampEnd<T>(ReadOnlySpan<T> span, int start, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            // Initially, start==len==0. If ClampStart trims all, start==len
+            Debug.Assert((uint)start <= span.Length);
+
+            int end = span.Length - 1;
+
+            for (; end >= start; end--)
+            {
+                if (!SequentialContains(trimElements, span[end]))
+                    break;
+            }
+
+            return end - start + 1;
+        }
+
+        /// <summary>
+        /// Scans for a specified item in the provided collection, returning true if found, else false.
+        /// Optimized for a small number of elements.
```

Comment:
Isn't IndexOfAny also optimized for a small number of elements, in particular 2 and 3?

Reply:
Because there's a ton of guards, type-checks (`byte`, `char`) and marshalling going on in there. I will bench it and report results

Reply:
> Because there's a ton of guards, type-checks (byte, char) and marshalling going on in there. I will bench it and report results

If it's insufficient for small sets, then I'd argue we should fix IndexOfAny to work better on small sets rather than write a custom version here, as small sets are a key use case for IndexOfAny.
cc: @benaadams

Reply:
type-checks and marshalling should all evaporate at Jit time

Reply:
Actually I am not sure that `IndexOfAny` is suitable. Or possibly I need coffee.
Let's say we are scanning a `RoS<int>` that is 100 elements long.
We want to remove any of `1, 2, 3`.
`IndexOfAny` tells us that it found a match at `[14]` (or `[44]` or `[99]`).
Well that's good to know, we know we don't need to slice since `[0]` wasn't a match.
But it failed to exit early - it should have stopped scanning after checking `[0]`.

Maybe there's a way to invert the search and get what we want?

Reply:
Ah, yeah you want a IndexOfNotAny; type thing...

Reply:
Ha. Yes I think you are right. Maybe that's a better name for the function, though it hurts my head to translate that :-)

Reply:
Added api request https://github.com/dotnet/corefx/issues/35594

Reply:
Isn't it actually just IndexOf / Contains we care about?  The loop is currently doing:
```C#
for (; start < span.Length; start++)
{
    if (!SequentialContains(trimElements, span[start]))
        break;
}
```
where SequentialContains is looking to see whether trimElements contains span[start].  How is that different from:
```C#
for (; start < span.Length; start++)
{
    if (!trimElements.Contains(span[start]))
        break;
}
```
?

Reply:
(But I see how an IndexNotOfAny would allow that loop to be replaced entirely.)

Reply:
Yes, that would be simpler if we could use `Contains`. 
Benching it, it's still more expensive as once again it's doing a lot of checks:
```
|        Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------- |----------:|----------:|----------:|------:|--------:|
| ViaSequential | 0.0009 ns | 0.0000 ns | 0.0000 ns |  1.00 |    0.00 |
| ViaIndexOfAny | 0.0016 ns | 0.0000 ns | 0.0000 ns |  1.90 |    0.06 |
|   ViaContains | 0.0010 ns | 0.0000 ns | 0.0000 ns |  1.17 |    0.03 | // <---
```

Reply:
> Benching it, it's still more expensive as once again it's doing a lot of checks

.0010ns vs .0009ns?  I'd prefer to take the .0001ns hit and avoid the duplicated code.

Reply:
Fair enough - the absolute values are more telling. Fixed.

Reply:
> Maybe there's a way to invert the search

I did need coffee